src/neuralnet/layer.cc

```diff
@@ -160,6 +160,200 @@ void DropoutLayer::ComputeGradient(const vector<SLayer>& srclayers)  {
+            for (int j = 0; j < vdim_; j++)
+              negsrc[i][j] = possrc[i][j];
+            /*Use the input samples for initializing the gibbs chain*/
+          hidden_data = dot(negsrc, weight);
+          is_first_iteration_bottom = false;
+        }
+        else {
+            Tensor<cpu, 2> hidden_data(hidden_data_.mutable_cpu_data(),
+                           Shape2(neg_batchsize_, hdim_));
+            Tensor<cpu, 2> negsrc(negsrc_.mutable_cpu_data(),
+                           Shape2(neg_batchsize_, vdim_));
+            Tensor<cpu, 2> weight(weight_->mutable_cpu_data(),
+                           Shape2(vdim_, hdim_));
+            Tensor<cpu, 1> bias(bias_->mutable_cpu_data(), Shape1(vdim_));
+            negsrc = dot(hidden_data, weight.T());
+            negsrc+=repmat(bias, neg_batchsize_);
+            negsrc = F<op::sigmoid>(negsrc);
+            for (int i = 0; i < neg_batchsize_; i++)
+              for (int j = 0; j < vdim_; j++)
+                negsrc[i][j] = static_cast<float>((rand() / static_cast<double>(RAND_MAX)) > negsrc[i][j] ? 0 : 1);
+            hidden_data = dot(negsrc, weight);
+        }
+  }
+}
+
+void DBMBottomLayer::ComputeGradient(const vector<SLayer>& srclayers) {
+  Tensor<cpu, 2> data(data_.mutable_cpu_data(), Shape2(batchsize_, hdim_));
+  Tensor<cpu, 2> hidden_data(hidden_data_.mutable_cpu_data(),
+              Shape2(neg_batchsize_, hdim_));
+  kPhase = true;
+  CHECK_EQ(srclayers[0]->data(this).count(), batchsize_*vdim_);
+  Tensor<cpu, 2> possrc(srclayers[0]->mutable_data(this)->mutable_cpu_data(),
+        Shape2(batchsize_, vdim_));
+  Tensor<cpu, 2> negsrc(negsrc_.mutable_cpu_data(),
+        Shape2(neg_batchsize_, vdim_));
+  Tensor<cpu, 2> gweight(weight_->mutable_cpu_grad(), Shape2(vdim_, hdim_));
+  Tensor<cpu, 1> gbias(bias_->mutable_cpu_grad(), Shape1(vdim_));
+  gbias = sum_rows(negsrc);
+  gbias-=sum_rows(possrc);
+  gweight = dot(negsrc.T(), hidden_data);
+  gweight-=dot(possrc.T(), data);
+  gbias*=scale_/(1.0f*batchsize_);
+  gweight*=scale_/(1.0f*batchsize_);
+}
+
+void DBMBottomLayer::ComputeLoss(Metric* perf) {
+  float loss = (0.0f);
+  kPhase = true;
+  CHECK_EQ(srclayers_[0]->data(this).count(), batchsize_*vdim_);
+  Tensor<cpu, 2> possrc(srclayers_[0]->mutable_data(this)->mutable_cpu_data(),
+         Shape2(batchsize_, vdim_));
+  Tensor<cpu, 2> data(data_.mutable_cpu_data(), Shape2(batchsize_, hdim_));
+  Tensor<cpu, 2> weight(weight_->mutable_cpu_data(), Shape2(vdim_, hdim_));
+  Tensor<cpu, 1> bias(bias_->mutable_cpu_data(), Shape1(vdim_));
+  Tensor<cpu, 2> reconstruct(Shape2(batchsize_, vdim_)); /*reconstruct error*/
+  AllocSpace(reconstruct);
+  reconstruct = dot(data, weight.T());
+  reconstruct+=repmat(bias, batchsize_);
+  reconstruct = F<op::sigmoid>(reconstruct);
+  for (int i = 0; i < batchsize_; i++)
+    for (int j = 0; j < vdim_; j++) {
+      loss += -(possrc[i][j]*log(reconstruct[i][j])+(1-possrc[i][j])*log(1-reconstruct[i][j]));
```

Comment:
replace to single loop and avoid using mshaow's [] operator. try to use mshadow's tensors.